jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java

```diff
@@ -457,6 +458,13 @@ else if (filled == 0)
         finally
         {
             setCurrentConnection(last);
+
+            // If handling has completed
+            if (_stream.get() == null && _retainableByteBuffer != null)
+            {
+                _retainableByteBuffer.release();
+                _retainableByteBuffer = null;
+            }
```

Comment:
This seems fishy.

The usual pattern is to release the buffer when a `fill()` returned `0` or `-1`, where we know we are not going to use the buffer.

Doing it in the finally is in a race condition, as when `filled==0` we set fill interest, and another thread could re-enter, use the current buffer that is actually being released, so not good.

The buffer should be properly handled in `fillRequestBuffer()`. What prompted you to add this?

Reply:
Yes it is fishy, hence I wrote:
> Whilst working on this, I noticed that onFillable could exit without releasing the request buffer. So I've added a failsafe release. However, it would be good to review the release logic of the class as part of the review of this PR

Releasing in the finally here is a fail safe I put in that was necessary since the buffer is not released if the endpoint is closed whilst there is a thread in handling.    This works and we could go forward with just that fix, but I am asking should we review/cleanup the whole releasing behaviour of this class in this PR (or is that for another PR?)

Reply:
Releasing in the finally is definitely racy.

Perhaps the best would be using `while (true)`, with a check inside for `endPoint.isOpen()`, and if not, release the buffer and break the loop.

Reply:
Also note that once _stream.get() == null, there will never be another thread entering here.  As I said, it is a fail-safe to protect against the `while(getEndPoint().isOpen()` above becoming an unreleasing noop if the endpoint is closed.